### PR TITLE
Make ShardedTensor ctor more inline with torch.Tensor ctor

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72164

torch.Tensor ctor creates an empty tensor and this PR makes
ShardedTensor on par with that.

In particular we remove TensorInitParams and instead always a create an empty
tensor and then fill it in for things like ones, zeros, full etc. This is
inline with torch.ones etc. as well since even for those APIs we first create
an empty tensor and then fill it out.

Differential Revision: [D33934603](https://our.internmc.facebook.com/intern/diff/D33934603/)

### CODE REVIEW

File: torch/nn/init.py [hunk 1]
```diff
@@ -190,6 +190,8 @@ def constant_(tensor: Tensor, val: float) -> Tensor:
         >>> w = torch.empty(3, 5)
         >>> nn.init.constant_(w, 0.3)
     """
+    if has_torch_function_variadic(tensor):
```
Comment (albanD) [inline torch/nn/init.py:193]: Isn't this already catching the `.fill_()` function? Why do we want this here?

File: torch/distributed/_sharded_tensor/ops/init.py [hunk 2]
```diff
@@ -86,3 +86,42 @@ def kaiming_uniform_(types, args=(), kwargs=None, pg=None):
     for shard in sharded_tensor.local_shards():
         torch.nn.init.kaiming_uniform_(shard.tensor, a=a, mode=mode, nonlinearity=nonlinearity)
     return sharded_tensor
+
+@sharded_op_impl(torch.nn.init.normal_)
+def normal_(types, args=(), kwargs=None, pg=None):
```
Comment (wanchaol) [inline torch/distributed/_sharded_tensor/ops/init.py:91]: it seems that we already got `normal_` in line 33?

**Review by wanchaol — APPROVED** [review]
looks good! One nit is that there seems to be a redefinition with `normal_`

File: torch/nn/init.py (hunk 1 above)
Comment (pritamdamania87) [reply torch/nn/init.py:193]: I think this calls tensor.fill_() under the hood and not torch.fill_(). I tried a simple test and it doesn't seem to work:

```
import torch

class MyTensor():

    def __torch_function__(self, func, types, args=(), kwargs=None):
        print(f'called: {func}')

t = MyTensor()
torch.nn.init.constant_(t, 0)
```

The error is:
```
'MyTensor' object has no attribute 'fill_'
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
 in <module>
      7 
      8 t = MyTensor()
----> 9 torch.nn.init.constant_(t, 0)
torch/nn/init.py in constant_(tensor, val)
    191         >>> nn.init.constant_(w, 0.3)
    192     """
--> 193     return _no_grad_fill_(tensor, val)
    194 
    195 
torch/nn/init.py in _no_grad_fill_(tensor, val)
     60 def _no_grad_fill_(tensor, val):
     61     with torch.no_grad():
---> 62         return tensor.fill_(val)
     63 
     64 
AttributeError: 'MyTensor' object has no attribute 'fill_'
```

Comment (albanD) [reply torch/nn/init.py:193]: Right if you're not using a Tensor subclass, you don;t get any method "for free" you have to add them all. That is expected.

**Pull request closed without merging**